[trace-view] Updated and extended documentation (#23661)
MystenLabs/sui: docs/content/references/ide/debugger.mdx

      :::

   1. The extension displays a filter prompt. Either type a filter string to target specific tests or leave the field blank to run all tests and press <kbd>Enter</kbd>.

      <img src="./images/filter_string.png" style="width: 50%; height: auto;" alt="Filter test string" />

   1. Find the generated traces in the `traces` directory.

      :::info
        If trace generation in the VSCode extension does not work for some reason, you can also generated traces by executing tests for your package with additional flags:
        ```bash
        sui move test  --trace-execution --disassemble
        ```
      :::

**II. Start debugging**
   1. Open the Move file containing your test.

   1. Select **Run** -> **Start Debugging** from the main menu.

      <img src="./images/start_debugging.png" style="width: 30%; height: auto;" alt="Start test debugging" />

   1. If the file has multiple tests, select the specific test from the dropdown menu.

      <img src="./images/test_selection.png" style="width: 50%; height: auto;" alt="Test selection" />

### Debugging on-chain transactions

Debugging an on-chain transaction is a two-step process:

> **I. Generate execution trace**

Reviewer: Using Roman numerals for lists aren't something in our style guide, so it looks a bit strange. Is there perhaps another way to format this list that does adhere to our style guide? https://docs.sui.io/style-guide#lists

> **I. Generate execution trace**

Author: Turned top-level list into "regular" numbered list and the inner one into a bulleted one

>   To generate a trace for a given transaction we need to know this transaction's digest (in the example below it's `0x42`).

Author: Applied the change manually after fixing a small typo:

You need to know a transaction's digest to generate a trace for it. For example, `0x42`.

#### Precise versions for all packages

To locate the right version of the source code for our user package, utilize a Sui explorer (e.g. [suiscan](https://suiscan.xyz)) and Sui's Move Package Registry ([MVR](https://www.moveregistry.com/)). This method will become more reliable with growing adoption of MVR in the Sui ecosystem.

Search for this package in a Sui explorer.

<img src="./images/deepbook_search.png" style="width: 50%; height: auto;" alt="Package search" />

You can see that the the package in question is `deepbook/core`. Look at the detailed information about this package available in the explorer.

<img src="./images/deepbook_explorer.png" style="width: 70%; height: auto;" alt="Package explorer" />

You can see that the package description in the explorer includes an MVR link. When you follow that link, you get to the MVR page containing a different kind of description for this package. The MVR description includes a link to the package's source code repository and information that this ID represent the 3rd version of this package.

![Deepbook MVR](./images/deepbook_mvr.png)

> While the source code repository for `deepbook/core` contains three different packages, the `pool` module in our PTB is only defined in the `deepbook` package.

Reviewer: ```suggestion
While the source code repository for `deepbook/core` contains three different packages, the `pool` module in the PTB is only defined in the `deepbook` package.
```

You can now proceed to building a correct version of this package to generated debugging metadata that must accompany package source code. It's a modified version of the process described [earlier](#metadata-generation). Assume that all commands are executed in `$ROOT_DIR` unless stated otherwise.

1. Clone the repository

    ```bash
    git clone https://github.com/MystenLabs/deepbookv3.git
    ```

2. Go to the `deepbook` package directory

    ```bash
    cd deepbookv3/packages/deepbook
    ```

3. Developers often tag code released at a given version so check if any tags are available

    ```bash
    git tag -l
    ```

>     As it happens, in this repository the tags indeed represent different package versions.

Reviewer: ```suggestion
    In this repository, the tags indeed represent different package versions.
```

    ```
    v1.0.0
    v2.0.0
    v3.0.0
    ```

>     You can then choose the exact version of the package to build.

Reviewer: ```suggestion
    Choose the exact version of the package to build.
```

    ```bash
    git checkout v3.0.0
    ```

>     If tags are not found you can proceed directly to the next step but with no guarantee that the most recent version of source code in the repository corresponds to the version of the package on chain (which may affect debugging capabilities).

Reviewer: ```suggestion
    If tags are not found, you can proceed directly to the next step, but there is no guarantee that the most recent version of source code in the repository corresponds to the version of the package on-chain, which may affect debugging capabilities.
```

> 4. Build the package and copying debugging metadata

Reviewer: ```suggestion
1. Build the package and copying the debugging metadata.
```

    ```bash
    sui move build
    ```

>     This will result in creating a `build` directory with a `deepbook` subdirectory in it containing all the relevant metadata, which you now copy to the right directory

Reviewer: ```suggestion
    This creates a `build` directory with a `deepbook` subdirectory in it containing all the relevant metadata, which you now copy to the right directory.
```

    ```bash
    cp -r ~/deepbookv3/packages/deepbook/build/deepbook ~/.replay/0xb29d83c26cdd2a64959263abbcfc4a6937f0c9fccaf98580ca56faded65be244/source
    ```

#### Precise versions for system packages

> If you cannot locate sources for the user package (`0xb29d83c26cdd2a64959263abbcfc4a6937f0c9fccaf98580ca56faded65be244` in our running example), you can still locate source code and its correct version for system packages that this user package depends on. This will allow you to at least view the source of system packages, such as Sui framework package, during the debugging session.

Reviewer: ```suggestion
If you cannot locate sources for a user package, you can still locate source code and its correct version for system packages that the user package depends on. This allows you to at least view the source of system packages during the debugging session.
```

> System packages in Sui (Sui framework package in particular), may only change at a very specific time, that is at a time when protocol version for Sui changes. These changes are tracked [here](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework-snapshot/manifest.json) - in this file you will find protocol version numbers (starting with 3) and corresponding git revisions that represent the exact version of Sui framework package's source code at a given protocol version. The remaining piece of information to locate is what protocol version was in effect when our user package was published on chain. You can accomplish this by running the following query against the mainnet's GraphQL [endpoint](https://graphql.mainnet.sui.io/graphql) where `address` parameter is the ID of our user package.

Reviewer: ```suggestion
System packages in Sui (the Sui framework package in particular), might only change when the protocol version for Sui changes. These changes [are tracked](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework-snapshot/manifest.json) and you will find protocol version numbers (starting with 3) and corresponding git revisions that represent the exact version of the Sui framework package's source code at a given protocol version. Then, locate which protocol version was in effect when the user package was published on-chain. Run the following query against the mainnet's GraphQL [endpoint](https://graphql.mainnet.sui.io/graphql) where the `address` parameter is the ID of the user package.
```

Author: I rewrote this a bit:

If you cannot locate sources for a user package, you can still locate source code and its correct version for system packages that the user package depends on. This allows you to at least view the source of system packages during the debugging session.

System packages in Sui (the Sui framework package in particular), might only change when the protocol version for Sui changes. These changes [are tracked](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework-snapshot/manifest.json) and include information about protocol version numbers (starting with 3) and corresponding git revisions that represent the exact version of the Sui framework package's source code at a given protocol version. This allows you to locate which protocol version was in effect when the user package was published on-chain.

Run the following query against the mainnet's GraphQL [endpoint](https://graphql.mainnet.sui.io/graphql) where the `address` parameter is the ID of the user package (`0xb29d83c26cdd2a64959263abbcfc4a6937f0c9fccaf98580ca56faded65be244` in this example).

```
{
    object(address: "0xb29d83c26cdd2a64959263abbcfc4a6937f0c9fccaf98580ca56faded65be244") {
        previousTransaction {
            effects {
                epoch {
                    protocolConfigs {
                        protocolVersion
                    }
                }
            }
        }
    }
}
```

> In the case of our running example, the result of this query looks as follows.

Reviewer: ```suggestion
The result of this query looks as follows.
```

Author: Applied this exact change manually

```
{
    "data": {
        "object": {
            "previousTransaction": {
                "effects": {
                    "epoch": {
                        "protocolConfigs": {
                            "protocolVersion": 84
                        }
                    }
                }
            }
        }
    }
}
```

> This means, that our user package was published at protocol version 84. The corresponding git revision from the protocol version tracking file is `25804c243d07dd73c0d199e7794383bd855cd436`. You can now follow the debugging metadata generation instructions for Sui framework code described [earlier](#metadata-generation). The only difference is that before building the framework package, you need choose the right git revision.

Reviewer: ```suggestion
In this example, the user package was published at protocol version 84. The corresponding git revision from the protocol version tracking file is `25804c243d07dd73c0d199e7794383bd855cd436`. Now, you can follow the debugging metadata generation instructions for Sui framework code. The only difference is that before building the framework package, you need choose the right git revision.
```

Author: We lost a link in the change which I think is useful to include:

In this example, the user package was published at protocol version 84. The corresponding git revision from the protocol version tracking file is `25804c243d07dd73c0d199e7794383bd855cd436`. Now, you can follow the debugging metadata generation [instructions](#metadata-generation) for Sui framework code. The only difference is that before building the framework package, you need choose the right git revision.
